### ISSUE TITLE
[AVX-512] Only `qword ptr` and `zmmword ptr` dereferences are folded into `vpand`?

### ISSUE THREAD

**Issue by Validark** [llvm/llvm-project#113262]
This code: ([Godbolt link](https://zig.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYgAzKVpMGoAF55gpJfWQE8Ayo3QBhVLQCuLBhICspJwAyeAyYAHKeAEaYxCAATBqkAA6oCoT2DK4eXr5JKWkCQSHhLFEx8VaYNnYCQgRMxASZnt5cftaYtum19QSFYZHRcQkKdQ1N2a1Wo73B/SWD8QCUVqjuxMjsHGgMIwDUI%2Bi7AKR6ACK7AAJ4LMkNEEexsQcPsYsnAEJHGgCCX9%2BYqluBF2VAYINQqAgTBAlwAah0iMQIAA2aS7dwADmWuwiMIAVBd4bYSBAuKR0VjFnCESTUeTMVSjgB2T4/Xbs3bETAENZgpjHWLI/YEdAAOlSLDFXMSmCYBBRaIiorxbz0rN%2BTNOfz%2BAKBILBEXqUJhhJpSLpFOxuN2BKJiIgsXplOpxPNaIZxxZfw5nO5vN2/IeQoO4uuUswMrlCvJSpVH21mu1P11JGBoJxTFMxpd9otDJj%2BNNrog7udRdzpcZXrZHK5POIfIFwZFocloulsvlFtjqvVzK1Px1gNT%2BpxAE9MNny7TKwWbdOkRinYsqQvo5bPeqfXX/YHBcKxRLw5Gu4rlb2EwPfsnhw1R4a8FO7TPl3PbWbScjl6vn27l5vvVrP0GwDJsD1bY9O3XHt40HRNBxvPV00NAB3J8PzzLE3zXPRHUtH8MMrACa3ZHcQL3ZtDzDdsIyg7tz1gjUryHJCDSYABHdDi0wq1C1/dcGQI7iiOZLcgPrRsg3Ao8aJPaCGLVS8OGWWhOB8XhvG4XhUE4AAtCx9lWdZMAFPQ9F4AhNGU5YEFlLAYggZYAGsQEkJlRUkHwMViPQMQxABOLg9CkJlDE4SReBYCQNASTStFIHSOF4BQQASSyOC0ZY4FgGBEBQVAbjoaJyEoNBCvoGJTAIYh3AYJy%2BDoAhohSiBcQy3gImCeox04HhSE65hiDHAB5CJtARXreDKthBGGhhaB69rSCwCJ3GAZwxFoFKtOWzAWGMYBxCW/AuU6AA3TBtvigEOncJrJvIQRKisww8AiYhutcLAXuq64Hou4gIhSTBTj2g7aGCUB2uWKgjGABRYTwTAUOGmVNL6/hBBEMR2CkGRBEUFR1CW3QySMEwQHMSwIYiFLIGWVBEmqHZOAAWmG3YACVKllJQADEmD2VnRmAbldlUDFkQAfVRVmIYYdxVF2VmWGQRJ3BOU5TAYAGIoSgHiDwLA6ccioqnSRwGBcNxmn0PxAlmYpSgkULklSZnxm8PQ/Dd/IGD6J3Bi4UL2k6Gppk9u2zYRLppgDgYYmDqYekj73k4aeP5kTpllgUIyNn0FS1I0l7Et2Kndmq2qnN2CBcEIEhTL0RYLKsldSFsph7MoZyQB8WJRR8SRg%2BRLguA0SWMR8AxVI4PWop8WLS84ZLUtIdLMtIbK8rwMgKAgDhWdBVBWfwBRRGIdBWbOsR3EwVnmDYBQOBesrEiK4hQlYTYK6rureCwPtEw9BWYKDHCUNwV1eCYHwIiI2%2BgGq0CasQVmcpfqrSalAxByDWbTUYAQLBNAkHRFPpgVawBWbBH4C9IhODggjBqmHBgKtuRdzlNCJamNhCiHEHjLhhM1AvVJtHJhltrZZC9kuB2RQE4SEdL7D2Ntsi%2BVyO7dImdnZcEdKHZm3QxhKMkSI3RcdHayK0enRoBj9BLhGD0DRQdXjry5JgeBCQUIfUSJNIuHB1KkDitpTgAQAiwgALK7AAJKc12IjZG0Ry4GT/jXOusDG4PDMuSVw5U4lpK4C3debcbJ2UGKbWe88%2B5LyWolVeaUCmkBcv3Qew8mSj3HpPaeYUODmT8cvJK%2BTobeNiCXSpK8%2BmZWWADVIDhJBAA%3D%3D%3D))

```zig
const std = @import("std");

export fn foo(a: @Vector(64, u8), b: *@Vector(1, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn bar(a: @Vector(64, u8), b: *@Vector(2, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn baz(a: @Vector(64, u8), b: *@Vector(4, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn bye(a: @Vector(64, u8), b: *@Vector(8, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn bai(a: @Vector(64, u8), b: *@Vector(16, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn baw(a: @Vector(64, u8), b: *@Vector(32, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}

export fn baq(a: @Vector(64, u8), b: *@Vector(64, u8)) @Vector(64, u8) {
    return a & std.simd.repeat(64, b.*);
}
```

Produces this for Zen 5 targets:

```asm
foo:
        vpbroadcastb    zmm1, byte ptr [rdi]
        vpandq  zmm0, zmm1, zmm0
        ret

bar:
        vpbroadcastw    zmm1, word ptr [rdi]
        vpandq  zmm0, zmm1, zmm0
        ret

baz:
        vbroadcastss    zmm1, dword ptr [rdi]
        vandps  zmm0, zmm1, zmm0
        ret

bye:
        vandpd  zmm0, zmm0, qword ptr [rdi]{1to8}
        ret

bai:
        vbroadcasti32x4 zmm1, xmmword ptr [rdi]
        vpandq  zmm0, zmm1, zmm0
        ret

baw:
        vbroadcasti64x4 zmm1, ymmword ptr [rdi]
        vpandq  zmm0, zmm1, zmm0
        ret

baq:
        vandps  zmm0, zmm0, zmmword ptr [rdi]
        ret
```

Why is it that only the `qword ptr` and `zmmword ptr` dereferences are folded into the `vpand`? Can't most or all of these be folded into the `vpand`?

**Comment by phoebewang** [comment]
Because it's not a full range memory fold but a broadcast fold. Only `baz` misses a fold. Others don't have instruction to support it.

**Comment by Validark** [comment]
> Because it's not a full range memory fold but a broadcast fold. Only `baz` misses a fold. Others don't have instruction to support it.

Is the `zmmword ptr` dereference the only supported option for things like `vpadd` and `vpcmpeq`?

**Comment by phoebewang** [comment]
> > Because it's not a full range memory fold but a broadcast fold. Only `baz` misses a fold. Others don't have instruction to support it.
> 
> Is the `zmmword ptr` dereference the only supported option for things like `vpadd` and `vpcmpeq`?

No, they also support broadcast fold, but only limited to element width.

**Comment by Validark** [comment]
> > > Because it's not a full range memory fold but a broadcast fold. Only `baz` misses a fold. Others don't have instruction to support it.
> > 
> > 
> > Is the `zmmword ptr` dereference the only supported option for things like `vpadd` and `vpcmpeq`?
> 
> No, they also support broadcast fold, but only limited to element width.

So this:

```asm
        vpbroadcastb    zmm1, byte ptr [rdi]
        vpcmpeqb        k0, zmm1, zmm0
```

Could be this?

```asm
        vpcmpeqb        k0, zmm0, byte ptr [rdi]
```

I think I see the potential discrepancy here. The ISA only has a few types of `vpand` granularity because it only makes a difference for pointer-dereference-folding. Otherwise, it's bit-by-bit, so larger subdivisions don't matter. Whereas, for things like `vpadd` and `vpcmpeq`, we have a separate instruction for byte-granularity, 2-byte-granularity, 4-byte-granularity, and 8-byte-granularity. In that case, we should be able to fold the pointer dereference into the instruction anytime the broadcast size matches the granularity at which the instruction operates. We wouldn't be able to do so if they differ, e.g. if we wanted to do a `dword broadcast` but feed that into a `vpcmpeqb`. Am I understanding correctly now?

**Comment by phoebewang** [comment]
No, unfortunately. Broadcast folding must be element width, but not all data type support broadcast folding. For vpcmpeq*, only 32 and 64 bit are supported according to SDM. I took a further look, seems we only support i32/i64/double/float/half/bfloat type of broadcast folding.
